Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.ipp

```diff
@@ -0,0 +1,135 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/Sycl/Seeding/CreateSeedsForGroupSycl.hpp"
+#include "Acts/Plugins/Sycl/Seeding/Seedfinder.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <utility>
+
+namespace Acts::Sycl {
+
+template <typename external_spacepoint_t>
+void Seedfinder<external_spacepoint_t>::init() {
+  // init m_config
+  m_config.highland = 13.6f * std::sqrt(m_config.radLengthPerSeed) *
+                      (1 + 0.038f * std::log(m_config.radLengthPerSeed));
+  float maxScatteringAngle = m_config.highland / m_config.minPt;
+  m_config.maxScatteringAngle2 = maxScatteringAngle * maxScatteringAngle;
+  m_config.pTPerHelixRadius = 300.f * m_config.bFieldInZ;
+  m_config.minHelixDiameter2 =
+      std::pow(m_config.minPt * 2 / m_config.pTPerHelixRadius, 2);
+  m_config.pT2perRadius =
+      std::pow(m_config.highland / m_config.pTPerHelixRadius, 2);
+
+  auto seedFilterConfig = m_config.seedFilter->getSeedFilterConfig();
+
+  // init m_deviceConfig
+  m_deviceConfig = Acts::Sycl::detail::DeviceSeedfinderConfig{
+      m_config.deltaRMin,
+      m_config.deltaRMax,
+      m_config.cotThetaMax,
+      m_config.collisionRegionMin,
+      m_config.collisionRegionMax,
+      m_config.maxScatteringAngle2,
+      m_config.sigmaScattering,
+      m_config.minHelixDiameter2,
+      m_config.pT2perRadius,
+      seedFilterConfig.deltaInvHelixDiameter,
+      seedFilterConfig.impactWeightFactor,
+      seedFilterConfig.deltaRMin,
+      seedFilterConfig.compatSeedWeight,
+      m_config.impactMax,
+      seedFilterConfig.compatSeedLimit,
+  };
+}
+
+template <typename external_spacepoint_t>
+Seedfinder<external_spacepoint_t>::Seedfinder(
+    Acts::SeedfinderConfig<external_spacepoint_t> config,
+    const Acts::Sycl::DeviceExperimentCuts& cuts)
+    : m_config(config), m_deviceCuts(cuts), m_wrappedQueue() {
+  init();
+}
+
+template <typename external_spacepoint_t>
+Seedfinder<external_spacepoint_t>::Seedfinder(
+    Acts::SeedfinderConfig<external_spacepoint_t> config,
+    const Acts::Sycl::DeviceExperimentCuts& cuts,
+    Acts::Sycl::QueueWrapper wrappedQueue)
+    : m_config(config),
+      m_deviceCuts(cuts),
+      m_wrappedQueue(std::move(wrappedQueue)) {
+  init();
```

Comment:
I'm giving the code one last read-through.

At this point I believe the code is in a reasonable shape to be merged in. I just want to point out a couple of things that I think could help you in your work later on.

I would not have used a separate `init()` function here. As that's a very [C\+\+03](https://en.wikipedia.org/wiki/C%2B%2B03) thing to do. :stuck_out_tongue: I would rather have done one of the following:
  - Have just one constructor, that receives a default-constructed `Acts::Sycl::QueueWrapper` object as its last argument. Which the client may or may not override.
  - Have 2 constructors defined. Put the full initialisation of the object into the constructor with the larger number of arguments, and have the constructor with fewer arguments call the other constructor. Like:

```c++
template <typename external_spacepoint_t>
Seedfinder<external_spacepoint_t>::Seedfinder(
    Acts::SeedfinderConfig<external_spacepoint_t> config,
    const Acts::Sycl::DeviceExperimentCuts& cuts)
    : Seedfinder(config, cuts, Acts::Sycl::QueueWrapper()) {}
```

Reply:
I wanted to use delegating constructors at first, but this didn't come to my mind. Neither did the first version. But I like that very much. I'll change this soon.